### ISSUE TITLE
Allow usage of the same attribute name in different branches within variant types

### ISSUE THREAD

**Issue by cryzed** [nim-lang/RFCs#19]
The following code snippet raises an error during compilation: `Error: redefinition of 'someAttribute'`. 

``` nim
  type
    MyKind = enum
      Kind1, Kind2, Kind3

    MyType = ref object
      case kind: MyKind
      of Kind1:
        someAttribute: string
      of Kind2:
        someAttribute: string
      of Kind3:
        someOtherAttribute: int
```

Since only a single branch, depending on `kind`, can be active at once, it would seem logical for this to work -- since no actual redefinition of `someAttribute` ever takes place. Currently I need to work around this by naming the fields for each branch differently and creating a custom umbrella accessor proc that has to check the object kind at runtime and return the correct field, which is associated with its own problems: #3628.

**Comment by Skaruts** [comment]
I'm a bit divided on the proposals, though I tend to lean on Araq's, though I agree with the concerns that it's not as readable. I'm not at all aware of what goes into implementing any of this, so my suggestion may be totally stupid, but what if you could use conditions inside the branches? 

```nim
Entry = ref object
    case kind
    of File,Dir:
        name: string
        modified: datetime
        attr: uint32
        group: uint32
        if File:
            size: uint64
            crc32: uint32
        if Dir:
            files: seq[Entry]
```

Just an example. It could be an `if` or a `when` or something else. Ideally you'd also be allowed to group discriminators in it.

**Comment by petabyteboy** [comment]
I think the initial proposal is least confusing, because this is what I was intuitively trying to do when I encountered this issue.

**Comment by krux02** [comment]
My proposal is, use this pattern and leave the language at it is:

```nim
type
  MyKind = enum
    Kind1, Kind2, Kind3

  Sub1 = object
    someAttribute: string

  Sub2 = object
    someAttribute: string

  Sub3 = object
    someOtherAttribute: int

  MyType = ref object
    case kind: MyKind
    of Kind1:
      sub1: Sub1
    of Kind2:
      sub2: Sub2
    of Kind3:
      sub3: Sub3

type
  EntryKind = enum
    File, Dir, Device

  EntryFile = object
    name: string
    modified: float64
    size: int64

  EntryDir = object
    name: string
    modified: float64

  EntryDevice = object
    name: string

  Entry = ref object
    case kind: EntryKind
    of File:
      file: EntryFile
    of Dir:
      dir: EntryDir
    of Device:
      device: EntryDevice
```

The advantage is, this won't silently break offsetof/sizeof calculation, which was hard to implement correctly. It already works. The only thing that needs to be done is, it should be documented.

**Comment by petabyteboy** [comment]
It does not work for my use case. I am parsing JSON-serialized data that looks like this:
```
[
                {
                        "type": "TypeA",
                        "address": 23
                        // has an integer as address
                },
                {
                        "type": "TypeB",
                        "address": "foo"
                        // has a string as address
                },
                {
                        "type": "TypeC",
                        "model": "bar"
                        // does not have an address attribute
                }
]
```

The only way to properly represent this in nim objects is by using the same attribute name in different variants.

**Comment by krux02** [comment]
@petabyteboy That looks to me more like an issue for the json (de)serializer.

**Comment by awr1** [comment]
The problem I keep thinking in the back of my head is "how are the fields actually positioned in the struct" for case sections with common fields?

Suppose we have this object:
```nim
type Object* = object
  case kind: Kind
  of File:
    modified: Date
    size: int64
    name: string
  of Directory:
    modified: Date
    name: string
  of Device:
    uuid: UUID
```
What is the underlying organization? Is it:
```c
union {
  struct {
    Date modified;
    int64 size;
    string name;
  } u_File;
  struct {
    Date modified;
    string name;
  } u_Directory;
  struct {
    UUID uuid;
  } u_Device;
};
```
Where we "just don't care" about the underlying layout?

Or is it more like:
```c
union {
  struct {
    Date modified;
    int64 size; // access prohibited whilst kind == Directory?
    string name;
  } u_File_Directory;
  struct {
    UUID uuid;
  } u_Device;
};
```
where we make an attempt to keep the location of a mutual fields unified?

**Comment by akbcode** [comment]
As long as it gets in, I don't mind which proposal is accepted. I think having macros for creating ADTs and pattern matching in the standard library would be nice.

**Comment by Skaruts** [comment]
I agree, having either one would be better than nothing. 

Why not both, though?

Opinions are divided and I can see why: the initial proposal seems more readable for simpler objects, while Araq's seems more readable (and more handy) for complex ones. As been said, one is more reliable to implement than the other, but from the user standpoint this discussion makes it seem like both are equally valid. Imo, user comfort should be a core principle in the production of any tool (and that seems to be true in Nim, in many cases), so this should weigh in the decision. 

So unless the hard option is actually unreasonably unreliable to implement, why not go for both? 

(Since Araq's proposal is easier, it could be implemented first, and then the other one could be implemented in a _"will be ready when it's ready"_ kind of way.)

**Comment by hiteshjasani** [comment]
I would love to have this feature and I expect I'll use variant types more often.

Do either of the proposals make it easier to have generic code do json roundtrip conversions for variants? Currently marshalling is handled generically for all variants, but you have to write custom code to unmarshall a json variant type. See [this gist](https://gist.github.com/hiteshjasani/1f8b962fa53e14316d61b73be8f4c7cd).

Either proposal is fine with me. Sure there are readability considerations on both sides. But the bigger question to me is how much additional custom code will I have to write to do normal conversions with each new variant type I add?

> It is better to have 100 functions operate on one data structure than 10 functions on 10 data structures. - Alan Perlis http://www.cs.yale.edu/homes/perlis-alan/quotes.html

**Comment by zedeus** [comment]
@hiteshjasani from Nim 1.2.0 your gist is wrong, it produces the expected result.

**Comment by Yardanico** [comment]
Today I also hit the same issue as @petabyteboy when considering using object variants for TDLib JSON object deserialization - objects in TDLib scheme can share some fields under the same name and type, so that even Araq's proposal would've worked fine.

I guess I'll have to generate a bit more sophisticated deserialization than just using ``json.to`` for now :)

**Comment by disruptek** [comment]
@Clyybber has a good proposal:

```nim
type
  NodeKind = enum
    foo
    bar
    bif
    baz
    bad

  Node = object
    case kind: NodeKind
    of foo, bar, baz:
      children: seq[Node]
    of bad, bif:
      dad: Node
    else: discard

    case kind
    of foo:
       additionalFieldForFoo: int
    of bar:
      onlyValidForBar: float
    else: discard

```

I like this because it's explicit, behaves as you might expect for a case
statement, and the semantics of when the fields appear are quite clear.

**Comment by Clyybber** [comment]
Ironically, I prefer Araq's proposal (https://github.com/nim-lang/RFCs/issues/19#issuecomment-173529871) over mine :D

**Comment by Yardanico** [comment]
@disruptek seems like this is the same as in https://github.com/nim-lang/RFCs/issues/19#issuecomment-440953864 :)

**Comment by tontinton** [comment]
Is this going to happen?

**Comment by Araq** [comment]
Yes. Eventually.

**Comment by disruptek** [comment]
I like the brevity of @Araq's version but c'mon, do we really want `case` syntax to vary here?  I still don't think that's a smart design choice, especially when I consider meta-programming these constructs.

**Comment by inv2004** [comment]
What do you think about @Araq's variant, which is unwrapped into initial variant, **but** it will raise an error if we there is the same name with different type?

I started to look at the RFC, and it is clear that @Araq version is more type-safe, but during the implementation I found that it is necessary 1) to change case-cover-check behavior 2) unwrap the a bit special case into separate variants to generate structs in the union, after it I decided that probably it is a bit more natural (and easier) to build structs like it is done already, but to raise error if type is not the same.

```nim
Entry = ref object
      case kind
      of File:
        name: string
        modified: datetime
        size: int64
      of Dir:
        name: string
        modified: time  // errormsg: "type for the same field cannot be redefined"
      of Device:
        name: string
```

Correct me if I am wrong - it is my first 60minutes in Nim's compiler code.

**Comment by Bulat-Ziganshin** [comment]
1. statically-typed languages are necessarily shoiuld have all fields of the same name having same type - otherwise `e.modified` type will be unknown at compile-tyime and you suddenly get a dynamically-typed language
2. https://github.com/nim-lang/RFCs/issues/19#issuecomment-517033520 mentioned another issue with my (i.e. classic) proposal - it leaves memory structure unobvious. should we have fixed position for fields with the same name in different cases? or we should obey order of the original definition?
3. I still insist on the pros/cons list I have given in https://github.com/nim-lang/RFCs/issues/19#issuecomment-405592412

**Comment by Bulat-Ziganshin** [comment]
Can we have both mine and Araq (https://github.com/nim-lang/RFCs/issues/19#issuecomment-173529871) proposals implemented? I bet his version would be useful for code generation and code clarity since it ensures that each field has its type written only once - it may be important for anonymous types. Also, can we allow both 
1. explicitly specify the discriminator type as in `case kind of KindType`
2. make it anonymous type as in `case kind` whose values are automatically deduced from the case labels?

As Pascal and ML languages shown, the second version is enough in most cases.

And I just realized that we can't have guarantee fixed field position with any proposal - sometimes we just don't have layouts for all cases that's compatible for same-named fields.

**Comment by treeform** [comment]
I fully support this. For me this is the biggest stumbling block with case variant objects and the reason I don't really use them. If this was fixed much of my code could be converted to them.

**Comment by inv2004** [comment]
@Bulat-Ziganshin
1. It is exactly why the errormsg in my example
2. I do not see generic solution for it:
```nim
type Object* = object
  case kind: Kind
  of File:
    modified: Date
    size: int64
    name: string
  of Directory:
    modified: Date
    name: string
    uuid: UUID
  of Device:
    uuid: UUID

The only generic solution is to have all available fields in struct, and restrict access, but I suppose it is not good idea and it brakes the main idea of unions. It would help for conversion from one kind to another, but it is prohibited also.

I think that the anonymous types is much longer story that they would have the same type-safe problem @Araq mentioned in the RFC until Nim has access to the fields(or the anonymous-types) without patter-matching guards.

**Comment by Araq** [comment]
> Ironically, I prefer Araq's proposal (#19 (comment)) over mine :D

I prefer @Clyybber 's proposal over mine.

**Comment by ErikSchierboom** [comment]
I agree with @petabyteboy:

> I think the initial proposal is least confusing, because this is what I was intuitively trying to do when I encountered this issue.

So from the principle of least surprise, I would be in favor of the original proposal. That being said, I'd be fine with Araq's proposal, while I'm less enthusiastic about @Clyybber's proposal.

**Comment by Q-Master** [comment]
Same for me, I think that initial proposal is good and intuitive while needs more text to write, but @Araq's proposal is ok too. Multiple case statements as in @Clyybber's one are too much confusing and least readable.

**Comment by Bulat-Ziganshin** [comment]
I just recalled what Pascal allows to specify common fields BEFORE the case part, which is enough for most use-cases. So our example can be written as:
```Nim
Entry = ref object
    name: string
    modified: datetime
    attr: uint32
    group: uint32
    case kind
    of File:
        size: uint64
        crc32: uint32
    of Dir:
        files: seq[Entry]
```

Moreover, Pascal also allows to make nested variable part, thus specifying common part for several variants
```Pascal
Type  
  MyRec = Record  
          X : Longint;  
          Case byte of  
            2 : (Y : Longint;  
                 case byte of  
                 3 : (Z : Longint);  
                 );  
          end;
```

**Comment by planetis-m** [comment]
The more I used object variants the more I prefer @krux02 [proposal](https://github.com/nim-lang/RFCs/issues/19#issuecomment-512871054). Note @mratsim has already [made](https://github.com/mratsim/trace-of-radiance/blob/master/trace_of_radiance/support/emulate_classes_with_ADTs.nim) an excellent module that does emulates classes/interfaces with this exact concept. Now the only thing missing, is to make a patty fork that works with these proposal.

**Comment by Araq** [comment]
https://github.com/nim-lang/RFCs/issues/368 is the winning proposal.
